barman/cloud_providers/google_cloud_storage.py

```diff
@@ -0,0 +1,324 @@
+# -*- coding: utf-8 -*-
+# © Copyright EnterpriseDB UK Limited 2018-2022
+#
+# This file is part of Barman.
+#
+# Barman is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# Barman is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Barman.  If not, see <http://www.gnu.org/licenses/>.
+
+import logging
+import os
+
+from barman.clients.cloud_compression import decompress_to_file
+from barman.cloud import CloudInterface, DecompressingStreamingIO, DEFAULT_DELIMITER
+
+try:
+    # Python 3.x
+    from urllib.parse import urlparse
+except ImportError:
+    # Python 2.x
+    from urlparse import urlparse
+
+try:
+    from google.cloud import storage
+    from google.api_core.exceptions import GoogleAPIError, Conflict
+except ImportError:
+    raise SystemExit("Missing required python module: google-cloud-storage")
+
+
+BASE_URL = "https://console.cloud.google.com/storage/browser/"
+
+
+class GoogleCloudInterface(CloudInterface):
+    """
+    This class implements CloudInterface for GCS with the scope of using JSON API
+
+    storage client documentation:  https://googleapis.dev/python/storage/latest/client.html
+    JSON API documentation: https://cloud.google.com/storage/docs/json_api/v1/objects
+    """
+
+    # This implementation uses JSON API . does not support real parallel upload.
+    # <<Within the JSON API, there is an unrelated type of upload also called a "multipart upload".>>
+    MAX_CHUNKS_PER_FILE = 1
+
+    # Since there is only on chunk  min size is teh same as max archive size
+    MIN_CHUNK_SIZE = 1 << 40
+
+    # https://cloud.google.com/storage/docs/json_api/v1/objects/insert
+    # Google json api  permit a maximum of 5TB per file
+    # This is a hard limit, while our upload procedure can go over the specified
+    # MAX_ARCHIVE_SIZE - so we set a maximum of 1TB per file
+    MAX_ARCHIVE_SIZE = 1 << 40
+
+    def __init__(self, url, jobs=1, encryption_scope=None, tags=None):
```

Comment:
The cli argument default value for [jobs](https://github.com/EnterpriseDB/barman/blob/2eef0d65d018d01f465b89fdc4d1b3e2038654e4/barman/clients/cloud_backup.py#L260) will always override the default here, so even though the upload is single threaded we're still forking two child processes by default.

Is it worth explicitly ignoring the jobs argument and instead always passing jobs=1 when creating the cloud interface at https://github.com/EnterpriseDB/barman/pull/391/files#diff-25794773c52ad1a8c49fa11547d2890473acd4aa332c87f2528c9c6d394406c0R103 ?

Reply:
Actually job value will no effect. MAX_CHUNKS_PER_FILE=1 imply files will never be split, although we have multiple processes. (job>1) .
I think your suggestion is a clean solution : pass job=1 in GoogleCloudInterface creation in __init__.py.